<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>RNA-seq: from counts to DE genes: First steps: filtering, visualisation, and basic normalisation</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/mb/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicons/mb/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicons/mb/favicon-16x16.png"><link rel="manifest" href="../favicons/mb/site.webmanifest"><link rel="mask-icon" href="../favicons/mb/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav mb"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="mb-logo" alt="RNA-seq: from counts to DE genes" src="../assets/images/mb-logo.svg"><span class="badge text-bg-warning">
          <abbr title="This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback.">
            <a href="https://docs.carpentries.org/resources/curriculum/lesson-life-cycle.html" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-triangle" style="border-radius: 5px"></i>
              Alpha
            </a>
            <span class="visually-hidden">This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../02-filtering-visualisation-normalisation.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav mb" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="RNA-seq: from counts to DE genes" src="../assets/images/mb-logo-sm.svg"></div>
    <div class="lesson-title-md">
      RNA-seq: from counts to DE genes
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            RNA-seq: from counts to DE genes
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  RNA-seq: from counts to DE genes
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 25%" class="percentage">
    25%
  </div>
  <div class="progress mb">
    <div class="progress-bar mb" role="progressbar" style="width: 25%" aria-valuenow="25" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../02-filtering-visualisation-normalisation.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-introduction.html">1. Introduction</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        2. First steps: filtering, visualisation, and basic normalisation
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#filtering">Filtering</a></li>
<li><a href="#visualisation">Visualisation</a></li>
<li><a href="#basic-normalisation">Basic normalisation</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-de-with-limma.html">3. Differential expression with limma</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-summary.html">4. Summary and exercise</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/01-introduction.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/03-de-with-limma.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/01-introduction.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Introduction
        </a>
        <a class="chapter-link float-end" href="../instructor/03-de-with-limma.html" rel="next">
          Next: Differential...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>First steps: filtering, visualisation, and basic normalisation</h1>
        <p>Last updated on 2025-10-21 |

        <a href="https://github.com/melbournebioinformatics/rna-seq-counts-to-genes/edit/main/episodes/02-filtering-visualisation-normalisation.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 30 minutes</p>

        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>FIXME</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>FIXME</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="filtering">Filtering<a class="anchor" aria-label="anchor" href="#filtering"></a></h2>
<hr class="half-width"><p>So, assuming your counts table has biological replicates in each
group (and any technical replicates have been taken care of), the first
thing we need to do is <em>filtering</em>. More precisely, removing
genes with little to no expression, because the former add “noise” while
the latter aren’t even relevant. There are various approaches to
filtering, but one of the simplest is to use the <code>limma</code>
function <code>filterByExpression</code>:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu">filterByExpr</span><span class="op">(</span><span class="va">dge</span>, group<span class="op">=</span><span class="va">group</span><span class="op">)</span> <span class="co"># logical vector indicating which rows of dge to keep</span></span>
<span><span class="fu">table</span><span class="op">(</span><span class="va">keep</span><span class="op">)</span> <span class="co"># 32069 genes will be removed; 14009 genes will be kept</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>keep
FALSE  TRUE
32069 14009 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dge</span> <span class="op">&lt;-</span> <span class="va">dge</span><span class="op">[</span><span class="va">keep</span>,,keep.lib.size<span class="op">=</span><span class="cn">FALSE</span><span class="op">]</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">dge</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span> <span class="co"># compare to pre-filtered data to see the difference</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   RA_1 RA_2 RA_3 TGFb_1 TGFb_2 TGFb_3 RA_TGFb_1 RA_TGFb_2 RA_TGFb_3 WT_1 WT_2
27   16   30   10     16     14     10        12        14        16    9   15
32  948 1158  880    940    880    943       656      1197       944 1123  967
33    9   21   19     12      4      6        17        17        18   12    9
34  609  711  577    752    688    647       476       837       688  717  709
36 1289 1617 1252   1427   1270   1216       993      1666      1440 1563 1445
44  873 1047  756   1110   1024   1009       819      1417      1144 1101  973
   WT_3
27   19
32 1375
33   24
34  846
36 1878
44 1347</code></pre>
</div>
<p>Roughly speaking, this strategy keeps genes that have a reasonable
number of counts in a worthwhile number samples, and removes them
otherwise (see function help for precise details).</p>
<p>Following filtering, you should check the <em>library size</em> of
each sample. The library size of a sample is simply the sum of the
counts in that library, i.e. <span class="math display">\[
\begin{align*}
\text{Library size for sample}\ j &amp;= \sum_{i} Y_{ij} \\
&amp; = \sum_{\substack{\text{all genes}}} \text{counts for sample}\ j
\end{align*}
\]</span> where <span class="math inline">\(Y_{ij}\)</span> is the count
for gene <span class="math inline">\(i\)</span> in sample <span class="math inline">\(j\)</span>. Here’s a plot of the library sizes for
our data:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lib.size.millons</span> <span class="op">&lt;-</span> <span class="fu">colSums</span><span class="op">(</span><span class="va">dge</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span> <span class="op">/</span> <span class="fl">1e6</span></span>
<span></span>
<span><span class="fu">barplot</span><span class="op">(</span><span class="va">lib.size.millons</span>, las<span class="op">=</span><span class="fl">2</span>, cex.names<span class="op">=</span><span class="fl">0.8</span>, main<span class="op">=</span><span class="st">"Library size (millons)"</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>h<span class="op">=</span><span class="fl">20</span>, lwd<span class="op">=</span><span class="fl">2</span>, col<span class="op">=</span><span class="st">"red2"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/02-filtering-visualisation-normalisation-rendered-unnamed-chunk-3-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Library size is an important quantity. One reason to check it is
because bigger is usually better: 5 million or less (not that good);
10-15 million (better); 20+ million (much better). This is because (1)
it’s a limited resource that needs to be “shared” between all the genes
(in proportion to their expression); and (2) bigger counts generally
means better statistical properties (e.g. power to detect DE). Another
reason to check is because it will help signpost what tools you should
use to analyse your data (more about this later).</p>
</section><section><h2 class="section-heading" id="visualisation">Visualisation<a class="anchor" aria-label="anchor" href="#visualisation"></a></h2>
<hr class="half-width"><p>The next thing to do is <em>visualise</em> your data. This allows you
to (1) see the landscape of biological differences between experimental
groups, and (2) assess whether the data is affected by technical
variation. Detecting technical variation is particularly important
because its presence leads to missed and/or false discoveries,
i.e. rubbish DE results. Thus, this step is crucial: analysing RNA-seq
data without visualisation is literally “flying blind”!</p>
<p>There are many visualisation methods. Each has different advantages
and show different aspects of the data—there is no single perfect
visualisation, so use several! Here we’ll discuss three:</p>
<ol style="list-style-type: decimal"><li><p><em>PCA plots</em>. These give a special low dimensional
representation of our (very) high dimensional data (i.e. where we’ve
taken thousands of measurements on each sample). A ``good” plot looks
like this: (1) replicates <em>cluster</em>; and (2) there’s some
<em>separation</em> of biological groups.</p></li>
<li><p><em>Boxplots</em>. These are very simple: you take all the
expression measurements in each sample and dump them into a boxplot, one
boxplot for each sample. A ``good” set of plots looks like this: the
boxplots are lined up, without much movement between them.</p></li>
<li><p><em>RLE plots</em>. These are the same as boxplots but with one
extra (crucial) step: subtracting the row median from each row before
dumping the observations into boxplots. This additional step removes the
between-gene variation, making the plot more sensitive than boxplots for
detecting technical variation. A ``good” plot looks like this: the
boxplot medians are lined up around zero and the boxes are roughly the
same size.</p></li>
</ol><p>If any of these plots don’t look ``good”, in the ways described, this
indicates the presence of technical variation.</p>
<p>Let’s make each plot then discuss them:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># log transform counts, note: 1/2 offset to avoid log transforming zeros</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu">log2</span><span class="op">(</span><span class="va">dge</span><span class="op">$</span><span class="va">counts</span> <span class="op">+</span> <span class="fl">1</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># PCA plot (nice ggplot version, potentially for publication)</span></span>
<span><span class="va">pca</span> <span class="op">&lt;-</span> <span class="fu">prcomp</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="va">y</span><span class="op">)</span>, center<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> </span>
<span><span class="fu">autoplot</span><span class="op">(</span><span class="va">pca</span>, <span class="fu">data.frame</span><span class="op">(</span>group<span class="op">=</span><span class="va">group</span><span class="op">)</span>, col<span class="op">=</span><span class="st">'group'</span>, size<span class="op">=</span><span class="fl">4</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"PCA plot"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: `aes_string()` was deprecated in ggplot2 3.0.0.
ℹ Please use tidy evaluation idioms with `aes()`.
ℹ See also `vignette("ggplot2-in-packages")` for more information.
ℹ The deprecated feature was likely used in the ggfortify package.
  Please report the issue at &lt;https://github.com/sinhrks/ggfortify/issues&gt;.
This warning is displayed once every 8 hours.
Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
generated.</code></pre>
</div>
<figure><img src="../fig/02-filtering-visualisation-normalisation-rendered-unnamed-chunk-4-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># boxplot and RLE plot (simple base R versions; they're generally not for publication)</span></span>
<span><span class="fu">par</span><span class="op">(</span>mfcol<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cols.group</span> <span class="op">&lt;-</span> <span class="va">group</span></span>
<span><span class="fu">levels</span><span class="op">(</span><span class="va">cols.group</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">hue_pal</span><span class="op">(</span><span class="op">)</span><span class="op">(</span><span class="fl">4</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">boxplot</span><span class="op">(</span><span class="va">y</span>, col<span class="op">=</span><span class="fu">as.character</span><span class="op">(</span><span class="va">cols.group</span><span class="op">)</span>, outline<span class="op">=</span><span class="cn">FALSE</span>, las<span class="op">=</span><span class="fl">2</span>, cex.axis<span class="op">=</span><span class="fl">0.6</span>, main<span class="op">=</span><span class="st">"Boxplots"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">RLE</span> <span class="op">&lt;-</span> <span class="va">y</span> <span class="op">-</span> <span class="fu">rowMedians</span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="fu">boxplot</span><span class="op">(</span><span class="va">RLE</span>, col<span class="op">=</span><span class="fu">as.character</span><span class="op">(</span><span class="va">cols.group</span><span class="op">)</span>, outline<span class="op">=</span><span class="cn">FALSE</span>, las<span class="op">=</span><span class="fl">2</span>, cex.axis<span class="op">=</span><span class="fl">0.6</span>, main<span class="op">=</span><span class="st">"RLE plot"</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>h<span class="op">=</span><span class="fl">0</span>, col<span class="op">=</span><span class="st">"grey20"</span>, lwd<span class="op">=</span><span class="fl">3</span>, lty<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/02-filtering-visualisation-normalisation-rendered-unnamed-chunk-5-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Note that we made these plots with log transformed counts, not the
counts themselves (the reason for this is technical: the log
transformation roughly removes the strong mean-variance relationship
inherent in RNA-seq data, something we don’t want driving our
visualisations).</p>
<p>For our data, the PCA plot looks okay; the replicates cluster
reasonably well and there’s separation of biological groups. The
boxplots and RLE plot, on the other hand, tell a different story (which
is why you should visualise your data multiple ways!). The boxplots are
wobbling around a little, which is a problem. Indeed, the RLE plot
reveals things are far from ideal. The boxplots and RLE plot clearly
indicate the presence of technical variation which needs to be removed
if we’re to obtain non-rubbish results. This is the topic we turn to
next.</p>
</section><section><h2 class="section-heading" id="basic-normalisation">Basic normalisation<a class="anchor" aria-label="anchor" href="#basic-normalisation"></a></h2>
<hr class="half-width"><p>Unfortunately, when doing RNA-seq analysis we can’t simply compare
counts between samples because there’s almost always technical
differences between them. To make samples properly comparable, the data
needs to be <em>normalised</em>: we need to remove or adjust for
technical differences between samples, leaving only biological
differences (the thing we’re interested in!). Failing to successfully
normalise your data can lead to significant missed and/or false
discoveries (rubbish results).</p>
<p>Two important technical differences between samples are (1) library
size differences and (2) compositional differences:</p>
<ol style="list-style-type: decimal"><li>
<em>Library size differences</em>. To see the problem here look at
Gene 3 in the made up table of counts:</li>
</ol><figure><img src="../fig/table-of-counts.png" style="width:70.0%" class="figure mx-auto d-block"></figure><p>Does Gene 3 show a biological or technical difference between the two
groups? Looks like a biological difference, but what if I told you the
library sizes for group B were 3 times bigger? In that case, the
difference looks more like an artifact of library size rather than a
true biological difference. Let’s look at the problem another way:</p>
<figure><img src="../fig/differences-lib-size.png" style="width:70.0%" class="figure mx-auto d-block"></figure><p>It should be clear from this that differences resulting from
different library sizes can be mistaken for biological differences. One
simple way to (roughly) normalise for library size differences is to
calculate <em>counts-per-million (CPM)</em>. If we again let <span class="math inline">\(Y_{ij}\)</span> be the count for gene <span class="math inline">\(i\)</span> in sample <span class="math inline">\(j\)</span>, and let <span class="math inline">\(N_j\)</span> be the library size for sample <span class="math inline">\(j\)</span>, then the CPM value for gene <span class="math inline">\(i\)</span> in sample <span class="math inline">\(j\)</span> is given by <span class="math display">\[
\text{CPM}_{ij} = \frac{Y_{ij}}{N_j} \times 10^6 =
\frac{\text{Count}}{\text{Library size}} \times 10^6,
\]</span> Because CPM values are (scaled) <em>proportions</em>,
i.e. “Count / Library size”, not absolute counts, the issue of library
size differences is no longer relevant.</p>
<ol start="2" style="list-style-type: decimal"><li>
<em>Compositional differences</em>. For each sample, the probability
of sequencing a transcript is proportional to its abundance—this is the
basis of how RNA-seq works (of course!). To see the potential problem
this fact creates, look at this figure:</li>
</ol><figure><img src="../fig/differences-composition.png" style="width:70.0%" class="figure mx-auto d-block"></figure><p>These two samples are the same (they have the same number of red,
yellow, blue, and green transcripts) except for the black transcript:
sample A has many more black transcripts. This difference in composition
means that even if the library sizes are the same, comparing raw counts
still gives the wrong impression. One popular approach for normalising
out these differences is the <code>TMM</code> method (from the
<code>edgeR</code> package).</p>
<p>Here’s how to adjust for library size and compositional differences
in our example:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dge</span> <span class="op">&lt;-</span> <span class="fu">calcNormFactors</span><span class="op">(</span><span class="va">dge</span>, method<span class="op">=</span><span class="st">"TMM"</span><span class="op">)</span> <span class="co"># adjustment factors for composition; stored in our container for later use</span></span>
<span><span class="va">CPM</span> <span class="op">&lt;-</span> <span class="fu">cpm</span><span class="op">(</span><span class="va">dge</span><span class="op">)</span> <span class="co"># adjustment factors for composition incorporated when calculating cpm values</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">CPM</span><span class="op">)</span> <span class="co"># compare to the raw counts</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     RA_1  RA_2   RA_3 TGFb_1 TGFb_2 TGFb_3 RA_TGFb_1 RA_TGFb_2 RA_TGFb_3
27  0.728  1.13  0.476  0.613  0.576   0.40     0.635     0.439     0.589
32 43.136 43.55 41.866 36.001 36.189  37.76    34.720    37.554    34.726
33  0.410  0.79  0.904  0.460  0.164   0.24     0.900     0.533     0.662
34 27.711 26.74 27.451 28.801 28.293  25.91    25.193    26.260    25.309
36 58.652 60.81 59.564 54.652 52.227  48.69    52.556    52.268    52.973
44 39.723 39.37 35.967 42.512 42.111  40.40    43.347    44.456    42.084
     WT_1   WT_2   WT_3
27  0.338  0.610  0.583
32 42.181 39.339 42.220
33  0.451  0.366  0.737
34 26.931 28.843 25.977
36 58.707 58.785 57.665
44 41.354 39.583 41.360</code></pre>
</div>
<p>Before we can proceed, however, we need to do something <em>very
important</em>: we need to <em>check</em> that the normalisation
methodology we’ve applied has actually worked! One way of doing this is
through visualisation. You want to see that the normalisation has (1)
removed technical (unwanted) variation, while (2) retaining (i.e. not
removing) the biological differences of interest. For example, here’s
what our data looks like after normalising for library size and
composition:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># log transform adjusted CPM values</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu">cpm</span><span class="op">(</span><span class="va">dge</span>, log<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># PCA plot (nice ggplot version, potentially for publication)</span></span>
<span><span class="va">pca</span> <span class="op">&lt;-</span> <span class="fu">prcomp</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="va">y</span><span class="op">)</span>, center<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> </span>
<span><span class="fu">autoplot</span><span class="op">(</span><span class="va">pca</span>, <span class="fu">data.frame</span><span class="op">(</span>group<span class="op">=</span><span class="va">group</span><span class="op">)</span>, col<span class="op">=</span><span class="st">'group'</span>, size<span class="op">=</span><span class="fl">4</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"PCA plot"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/02-filtering-visualisation-normalisation-rendered-unnamed-chunk-7-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># boxplot and RLE plot (simple base R versions; they're generally not for publication)</span></span>
<span><span class="fu">par</span><span class="op">(</span>mfcol<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cols.group</span> <span class="op">&lt;-</span> <span class="va">group</span></span>
<span><span class="fu">levels</span><span class="op">(</span><span class="va">cols.group</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">hue_pal</span><span class="op">(</span><span class="op">)</span><span class="op">(</span><span class="fl">4</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">boxplot</span><span class="op">(</span><span class="va">y</span>, col<span class="op">=</span><span class="fu">as.character</span><span class="op">(</span><span class="va">cols.group</span><span class="op">)</span>, outline<span class="op">=</span><span class="cn">FALSE</span>, las<span class="op">=</span><span class="fl">2</span>, cex.axis<span class="op">=</span><span class="fl">0.6</span>, main<span class="op">=</span><span class="st">"Boxplots"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">RLE</span> <span class="op">&lt;-</span> <span class="va">y</span> <span class="op">-</span> <span class="fu">rowMedians</span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="fu">boxplot</span><span class="op">(</span><span class="va">RLE</span>, col<span class="op">=</span><span class="fu">as.character</span><span class="op">(</span><span class="va">cols.group</span><span class="op">)</span>, outline<span class="op">=</span><span class="cn">FALSE</span>, las<span class="op">=</span><span class="fl">2</span>, cex.axis<span class="op">=</span><span class="fl">0.6</span>, main<span class="op">=</span><span class="st">"RLE plot"</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>h<span class="op">=</span><span class="fl">0</span>, col<span class="op">=</span><span class="st">"grey20"</span>, lwd<span class="op">=</span><span class="fl">3</span>, lty<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/02-filtering-visualisation-normalisation-rendered-unnamed-chunk-8-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Note that we can now use the cpm function to calculate logged data.
The PCA plot looks better: the replicates are clustered more tightly
(technical variation has been removed), and the biological groups are
more separated (biological differences between groups are maintained,
actually improved). Moreover, the boxplots and RLE plot are looking much
better: the boxplots are lined up, and the RLE plot medians are centered
on zero and the boxes are roughly similar in size.</p>
<p>WARNING!…library size and composition differences aren’t the only
kinds of technical differences that can arise between samples:</p>
<ul><li>
<p>Another way they can arise is when different samples are prepared
differently, e.g.</p>
<ul><li>with different protocols, or</li>
<li>in different labs, or</li>
<li>on different dates, or</li>
<li>by different people, or</li>
<li>etc.</li>
</ul><p>These differences can lead to so-called “batch effects”. For example,
consider these PCA plots of a real data set:</p>
</li>
</ul><figure><img src="../fig/batch-effect.png" style="width:80.0%" class="figure mx-auto d-block"></figure><p>These are both the “same” PCA plot, but one is coloured by biological
condition and the other is coloured by batch. The major factor driving
separation (along PC1) is batch, not biological condition; consequently,
replicates are not clustering by group, and the groups aren’t
separating. Batch effects like these can only be corrected with more
sophisticated analysis approaches, a topic outside the scope of this
workshop.</p>
<ul><li>Moreover, batch effects aren’t the only way problems can arise
(beyond library size and composition). For example, consider this PCA
plot of another real data set (coloured by biological condition):</li>
</ul><figure><img src="../fig/bad-PCA.png" style="width:40.0%" class="figure mx-auto d-block"></figure><p>Again, the replicates aren’t really clustering and the groups aren’t
really separating. The problem here, however, isn’t batch effects—it’s
something much more subtle.</p>
<p>The bottom line is this: if, after normalising for library size and
composition, your visualisations don’t look reasonably “good”, in the
ways discussed above, the results you’ll get from proceeding with a DE
analysis will likely contain missed and/or false discoveries—you’ll
likely get rubbish results.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul><li>FIXME</li>
</ul></div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/01-introduction.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/03-de-with-limma.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/01-introduction.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Introduction
        </a>
        <a class="chapter-link float-end" href="../instructor/03-de-with-limma.html" rel="next">
          Next: Differential...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/melbournebioinformatics/rna-seq-counts-to-genes/edit/main/episodes/02-filtering-visualisation-normalisation.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/melbournebioinformatics/rna-seq-counts-to-genes/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/melbournebioinformatics/rna-seq-counts-to-genes/" class="external-link">Source</a></p>
				<p><a href="https://github.com/melbournebioinformatics/rna-seq-counts-to-genes/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:vinicius.salazar%20%5B%20at%20%5D%20unimelb.edu.au">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors.</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a>.</p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.1" class="external-link">sandpaper (0.17.1)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/melbournebioinformatics/uom-varnish/tree/main" class="external-link">varnish (1.0.7)</a>.</p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "melbournebioinformatics.github.io/uom-varnish/instructor/02-filtering-visualisation-normalisation.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "transcriptomics, differential expression, high dimensional, multivariate",
  "name": "First steps: filtering, visualisation, and basic normalisation",
  "creativeWorkStatus": "active",
  "url": "melbournebioinformatics.github.io/uom-varnish/instructor/02-filtering-visualisation-normalisation.html",
  "identifier": "melbournebioinformatics.github.io/uom-varnish/instructor/02-filtering-visualisation-normalisation.html",
  "dateCreated": "2025-10-26",
  "dateModified": "2025-10-21",
  "datePublished": "2025-10-21"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

